File: Assets/Import_out/Uzil/API/BlockPage.cs
using UnityEngine;

using Uzil;
using Uzil.BlockPage;

namespace UZAPI {

public class BlockPage {

	/*======================================Constructor==========================================*/

	/*=====================================Static Members========================================*/

	/*=====================================Static Funciton=======================================*/

	/*=========================================Members===========================================*/

	/*========================================Components=========================================*/

	/*==========================================Event============================================*/

	/*======================================Unity Function=======================================*/

	/*========================================Interface==========================================*/

	/** 更新排版 */
	public static void UpdateLayout (string pageID) {
		BlockPageObj page = BlockPageUtil.GetPage(pageID);
		// if (page == null) return;
		page.UpdateLayout();
	}

	/** 新增 區塊 */
	public static void AddBlock (string pageID, string blockID, string blockType, string blockJson) {
		BlockPageObj page = BlockPageUtil.GetPage(pageID);
		if (page == null) return;

		DictSO data = DictSO.Json(blockJson);

		BlockObj block = BlockPageUtil.CreateBlock(blockType);
		page.AddBlock(blockID, block);

		block.SetData(data);
	}

	/** 新增 區塊 */
	public static void InsertBlock (string pageID, int rowIdx, string blockID, string blockType, string blockJson) {
		BlockPageObj page = BlockPageUtil.GetPage(pageID);
		if (page == null) return;

		DictSO data = DictSO.Json(blockJson);

		BlockObj block = BlockPageUtil.CreateBlock(blockType);
		page.InsertBlock(rowIdx, blockID, block);

		block.SetData(data);
	}

	/** 設置 區塊 */
	
	public static void SetBlock (string pageID, string blockID, string blockJson) {
		BlockPageObj page = BlockPageUtil.GetPage(pageID);
		if (page == null) return;

		BlockObj block = page.GetBlock(blockID);
		if (block == null) return;

		DictSO data = DictSO.Json(blockJson);
		block.SetData(data);
	}

	/** 移除 區塊 */
	public static void RemoveBlock (string pageID, string blockID) {
		BlockPageObj page = BlockPageUtil.GetPage(pageID);
		if (page == null) return;

		page.RemoveBlock(blockID);
	}

	/** 清空 區塊 */
	public static void ClearBlocks (string pageID) {
		BlockPageObj page = BlockPageUtil.GetPage(pageID);
		if (page == null) return;

		page.ClearBlocks();	
	}

	/** 移除 行 */
	public static void RemoveRow (string pageID, int rowIdx) {
		BlockPageObj page = BlockPageUtil.GetPage(pageID);
		if (page == null) return;

		page.RemoveRow(rowIdx);
	}

	/** 移除 所有 */
	public static void Clear (string pageID) {
		BlockPageObj page = BlockPageUtil.GetPage(pageID);
		if (page == null) return;

		page.Clear();
	}



	/*=====================================Public Function=======================================*/

	/*===================================Protected Function======================================*/

	/*====================================Private Function=======================================*/

}

}

File: AppData/Mods/BUILT-IN/_lua/LuaScript/Uzil/BlockPage.lua
local this = {};

--[[ 區塊類型 ]]
this.BlockType = {
	["empty"] = "Empty",
	["text"] = "Text",
	["image"] = "Image",
	["button"] = "Button",
	["toggle"] = "Toggle",
	["inputField"] = "InputField",
	["select"] = "Select",
	["switch"] = "Switch",
	["slider"] = "Slider",
};

--[[ 取得實例 ]] -- table
function this.inst(pageID)

	local instance = {};

	instance.pageID = pageID;

	--[[ 新增 區塊 ]] -- void
	instance.addBlock = function (blockID, blockType, data)
		this.addBlock(instance.pageID, blockID, blockType, data);
	end

	--[[ 新增 區塊 ]]
	instance.insertBlock = function (rowIdx, blockID, blockType, data)
		this.insertBlock(instance.pageID, rowIdx, blockID, blockType, data);
	end

	--[[ 設置 區塊 ]]
	instance.setBlock = function (blockID, data)
		this.setBlock(instance.pageID, blockID, data);
	end

	--[[ 移除 區塊 ]]
	instance.removeBlock = function (blockID)
		this.removeBlock(instance.pageID, blockID);
	end

	--[[ 清除 所有區塊 ]]
	instance.clearBlocks = function ()
		this.clearBlocks(instance.pageID);
	end

	--[[ 移除 行 ]]
	instance.removeRow = function (rowIdx)
		this.removeRow(instance.pageID, rowIdx);
	end

	--[[ 更新 排版 ]]
	instance.updateLayout = function ()
		this.updateLayout(instance.pageID);
	end

	return instance;
end

--[[ 新增 區塊 ]] -- void
function this.addBlock (pageID, blockID, blockType, data)
	UZAPI.BlockPage.AddBlock(pageID, blockID, blockType, Json.encode(data));
end

--[[ 新增 區塊 ]]
function this.insertBlock (pageID, rowIdx, blockID, blockType, data)
	UZAPI.BlockPage.InsertBlock(pageID, rowIdx, blockID, blockType, Json.encode(data));
end

--[[ 設置 區塊 ]]
function this.setBlock (pageID, blockID, data)
	UZAPI.BlockPage.SetBlock(pageID, blockID, Json.encode(data));
end

--[[ 移除 區塊 ]]
function this.removeBlock (pageID, blockID)
	UZAPI.BlockPage.RemoveBlock(pageID, blockID);
end

--[[ 清除 所有區塊 ]]
function this.clearBlocks (pageID)
	UZAPI.BlockPage.ClearBlocks(pageID);
end

--[[ 移除 行 ]]
function this.removeRow (pageID, rowIdx)
	UZAPI.BlockPage.RemoveRow(pageID, rowIdx);
end

--[[ 更新 排版 ]]
function this.updateLayout (pageID)
	UZAPI.BlockPage.UpdateLayout(pageID);
end

return this;